File: Upcoder/Lập trình nâng cao/1.Quá tải toán tử/Đạo hàm hàm đa thức/DAOHAM.cs
using System;
using System.Collections.Generic;
namespace DaoHam
{
    class DaThuc {
        public int n;
        public List<int> A;

        public DaThuc() {
            n = 0;
            A = new List<int>();
        }

        // Hàm nhập
        public static DaThuc FromInput() {
            DaThuc dt = new DaThuc();
            dt.n = int.Parse(Console.ReadLine());
            string[] input = Console.ReadLine().Split();
            dt.A = new List<int>(Array.ConvertAll(input, int.Parse));
            return dt;
        }

        // Hàm xuất
        public override string ToString() {
            bool isFirst = true;
            string result = "";
            for (int i = 0; i <= n; i++) {
                int coef = A[i];
                int exp = n - i;
                if (coef == 0) continue;
                if (!isFirst) {
                    result += coef > 0 ? "+" : "";
                } else {
                    isFirst = false;
                }
                if (Math.Abs(coef) != 1 || exp == 0) result += coef;
                else if (coef == -1) result += "-";
                if (exp > 0) result += "x";
                if (exp > 1) result += "^" + exp;
            }
            return isFirst ? "0" : result;
        }

        // Hàm tính đạo hàm
        public DaThuc DaoHam() {
            DaThuc res = new DaThuc();
            if (n == 0) {
                res.n = 0;
                res.A.Add(0);
                return res;
            }
            res.n = n - 1;
            for (int i = 0; i < n; i++) {
                res.A.Add(A[i] * (n - i));
            }
            return res;
        }
    }

    class Program {
        static void Main() {
            DaThuc p = DaThuc.FromInput();
            Console.WriteLine(p);
            DaThuc p1 = p.DaoHam();
            Console.WriteLine(p1);
            DaThuc p2 = p1.DaoHam();
            Console.Write(p2);
        }
    }
}

File: Upcoder/Lập trình nâng cao/1.Quá tải toán tử/Đạo hàm hàm đa thức/DAOHAM.py
class DaThuc:
    def __init__(self):
        self.n = 0
        self.A = []

    # Hàm nhập
    @staticmethod
    def from_input():
        dt = DaThuc()
        dt.n = int(input())
        dt.A = list(map(int, input().split()))
        return dt

    # Hàm xuất
    def __str__(self):
        is_first = True
        result = ""
        for i in range(self.n + 1):
            coef = self.A[i]
            exp = self.n - i
            if coef == 0:
                continue
            if not is_first:
                result += "+" if coef > 0 else ""
            else:
                is_first = False
            if abs(coef) != 1 or exp == 0:
                result += str(coef)
            elif coef == -1:
                result += "-"
            if exp > 0:
                result += "x"
            if exp > 1:
                result += "^" + str(exp)
        return "0" if is_first else result

    # Hàm tính đạo hàm
    def dao_ham(self):
        res = DaThuc()
        if self.n == 0:
            res.n = 0
            res.A.append(0)
            return res
        res.n = self.n - 1
        for i in range(self.n):
            res.A.append(self.A[i] * (self.n - i))
        return res


if __name__ == "__main__":
    p = DaThuc.from_input()
    print(p)
    p1 = p.dao_ham()
    print(p1)
    p2 = p1.dao_ham()
    print(p2, end="")
